numpy/core/src/private/ufunc_override.c

```diff
@@ -134,6 +134,7 @@ PyUFunc_WithOverride(PyObject *args, PyObject *kwds,
                 goto fail;
             }
             if (with_override != NULL) {
+                Py_INCREF(obj);
```

Comment:
It was just a question of where the `Py_INCREF` happens -- in @eric-wieser's initial attempt it happened just before being inserted in a tuple. Also, I thought it was annoying to pass back new references to the `__array_ufunc__` methods and borrowed references to `obj`. But I can change it if you think it is better.